### PR TITLE
feat: Guild Rename service

### PR DESCRIPTION
**Pull request by Nyeriah**
Wip

### CODE REVIEW

**Comment by Nyeriah** [comment]
Validation added. Tests performed:
- Non-guild masters can't change guild names
- Guild masters can change guild names

Anything else I might be missing?

**Comment by gareth-charles** [comment]
Re: validation.

Game server [imposes several restrictions](https://github.com/azerothcore/azerothcore-wotlk/blob/master/src/server/game/Globals/ObjectMgr.cpp#L8115), eg.

**Min length**:
- [defined by config](https://github.com/azerothcore/azerothcore-wotlk/blob/78bdc898813910c79c3a9c67f2bf9a0f1abef221/src/server/game/World/World.cpp#L755), with fallback value of 2.

**Max length**:
-  [hard coded](https://github.com/azerothcore/azerothcore-wotlk/blob/master/src/server/game/Guilds/Guild.cpp#L1187) max name length is 24.
- [additional verification here](https://github.com/azerothcore/azerothcore-wotlk/blob/09df444187066ce9ecfaa31cede12905c73b94fb/src/server/game/Globals/ObjectMgr.h#L685), where it is noted as being the hard coded length in the client.

Additional requirements based on [server language settings](https://github.com/azerothcore/azerothcore-wotlk/blob/master/src/server/game/Globals/ObjectMgr.cpp#L8054).

I think your best choice would be to ask the game server to validate the name for you, so we don't have to reproduce the validation on both places.

File: srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/GuildChange.php [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        global $product;
+        if ($product->get_sku() != "guild-rename") {
+            return;
+        }
+
+        $current_user = wp_get_current_user();
+
+        if ($current_user) {
+            ?>
+            <p>You must select the guild master character.</p>
```
Comment (gareth-charles) [inline srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/GuildChange.php:29]:
```suggestion
            <p>Please select the guild master of the guild whose name you wish to change.</p>
```
Optional text edit.

File: srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/GuildChange.php [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            <?php
+            FieldElements::charList($current_user->user_login);
+            ?>
+            <br>
+            <br>
+            <label for="acore_new_guild_name">New name:</label>
```
Comment (gareth-charles) [inline srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/GuildChange.php:35]:
```suggestion
            <label for="acore_new_guild_name">Please enter a new guild name:</label>
```
Optional text edit.

File: srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/GuildChange.php [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+            <input type="text" maxlength="20" id="acore_new_guild_name" class="acore_new_guild_name" name="acore_new_guild_name">
```
Comment (gareth-charles) [inline srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/GuildChange.php:36]:
```suggestion
            <input type="text" maxlength="24" id="acore_new_guild_name" class="acore_new_guild_name" name="acore_new_guild_name">
```
To match values found in server code.

File: srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/GuildChange.php [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+            <br>
+            <br>
+            <?php
+        }
+    }
+
+    // 3) SAVE INTO ITEM DATA
```
Comment (gareth-charles) [inline srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/GuildChange.php:43]: Very minor, but your comment counter goes 1), 3), 4), ... (ie. 2 is skipped)

File: srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Manager/Soap/GuildService.php [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+namespace ACore\Manager\Soap;
+
+use ACore\Manager\Soap\AcoreSoapTrait;
+
+class GuildService {
+
+    use AcoreSoapTrait;
+
+    public function guildRename($oldname, $newname) {
+        return $this->executeCommand(".guild rename \"$oldname\" \"$newname\"");
```
Comment (gareth-charles) [inline srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Manager/Soap/GuildService.php:12]: If this is the standard format of passing commands to the game server, I recommend filtering out `"` characters from inputs on php side.

File: srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/GuildChange.php (hunk 4 above)
Comment (Nyeriah) [reply srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/GuildChange.php:43]: It’s the same on every other service file, not really sure why

Comment (Helias) [reply srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/GuildChange.php:43]: to be honest I don't remember why 😓 , we can even remove the numbers and leave only the comments, that is the important part

File: srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/CartValidation.php [hunk 6]
```diff
@@ -15,6 +15,7 @@ class CartValidation extends \ACore\Lib\WpClass {
         "char-restore-delete" => ['acore_char_sel'],
         "char-transfer-sku" => ['acore_char_sel', 'acore_dest_account'],
         "itemsend" => ['acore_char_sel'],
+        "guild-rename" => ['acore_char_sel', 'acore_new_guild_name'],
```
Comment (Helias) [inline srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/CartValidation.php:18]: when you add a new service, add the SKU in the documentation https://github.com/azerothcore/acore-cms/blob/master/docs/configure-cms.md

File: srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/GuildChange.php [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        $logs = new \WC_Logger();
+
+        $order = new \WC_Order($order_id);
+        $items = $order->get_items();
+
+        $soap = ACoreServices::I()->getServerSoap();
+
+        foreach ($items as $item) {
+            if ($item["acore_item_sku"]) {
+                if ($item["acore_item_sku"] == "guild-rename") {
```
Comment (Helias) [inline srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/GuildChange.php:110]: minor: you could merge the two if conditions in one if

File: srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/GuildChange.php [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+            $soap = $WoWSrv->getGuildSoap();
+            
+            foreach ($items as $item) {
+                if (isset($item["acore_item_sku"])) {
+                    if ($item["acore_item_sku"] == "guild-rename") {
+                        $guildName = $WoWSrv->getGuildNameByLeader($item["acore_char_sel"]);
+
+                        if (empty($guildName))
+                        {
+                            throw new \Exception("The character is not the guild master.");
```
Comment (Helias) [inline srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/GuildChange.php:138]: can this be checked before? during the char selection?

Comment (Nyeriah) [reply srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/GuildChange.php:138]: I believe it's checked in both places already, unless I'm mistaken, or I didn't get what you mean specifically

File: srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/GuildChange.php [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        if ($current_user) {
+            ?>
+            <p>Please select the guild master of the guild whose name you wish to change.</p>
+            <?php
+            FieldElements::charList($current_user->user_login);
+            ?>
+            <br>
+            <br>
+            <label for="acore_new_guild_name">Please enter a new guild name:</label>
+            <input type="text" maxlength="24" id="acore_new_guild_name" class="acore_new_guild_name" name="acore_new_guild_name">
```
Comment (Helias) [inline srv/wordpress/wp-content/plugins/acore-wp-plugin/src/Hooks/WooCommerce/GuildChange.php:36]:
```suggestion
            <input type="text" maxlength="24" id="acore_new_guild_name" class="acore_new_guild_name" name="acore_new_guild_name" style="width: 300px;">
```

**Pull request merged**